arbnode/espresso_caff_node.go

```diff
@@ -0,0 +1,536 @@
+package arbnode
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	espressoClient "github.com/EspressoSystems/espresso-sequencer-go/client"
+	flag "github.com/spf13/pflag"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/offchainlabs/nitro/arbos"
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/espressostreamer"
+	"github.com/offchainlabs/nitro/execution/gethexec"
+	"github.com/offchainlabs/nitro/solgen/go/bridgegen"
+	"github.com/offchainlabs/nitro/util/dbutil"
+	"github.com/offchainlabs/nitro/util/headerreader"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+var (
+	DelayedFetcherCurrentL1BlockKey = []byte("delayedFetcherCurrentL1Block")
+	// To not to mess with the existing schema, we use another prefix
+	DelayedMessagePrefix = []byte("x")
+)
+
+type DelayedMessageFetcherInterface interface {
+	getDelayedMessage(index uint64) (*arbostypes.L1IncomingMessage, error)
+	reset(seqNum uint64)
+  getDelayedMessageCountAtBlock(blockNumber uint64) (uint64, error)
+}
+
+type DelayedMessageFetcher struct {
+	fromBlock     uint64
+	delayedBridge *DelayedBridge
+	l1Reader      *ethclient.Client
+
+	db ethdb.Database
+}
+
+func NewDelayedMessageFetcher(delayedBridge *DelayedBridge, l1Reader *ethclient.Client, db ethdb.Database) *DelayedMessageFetcher {
+	var fromBlock uint64
+	fromBlock, err := readCurrentL1BlockFromDb(db)
+	if err != nil {
+		log.Crit("failed to read l1 block from db", "err", err)
+		return nil
+	}
+
+	if fromBlock == 0 {
+		fromBlock = delayedBridge.fromBlock
+	}
+
+	return &DelayedMessageFetcher{
+		fromBlock:     fromBlock,
+		delayedBridge: delayedBridge,
+		l1Reader:      l1Reader,
+		db:            db,
+	}
+}
+// getDelayedMessageCountAtBlock is a wrapper function for the delayedBridge.GetMessageCount function. This allows users of the DelayedMessageFetcher
+// to query for the message count at a block. 
+func (f *DelayedMessageFetcher) getDelayedMessageCountAtBlock(blockNumber uint64) (uint64, error){
+  count, err := f.delayedBridge.GetMessageCount(nil, new(big.Int).SetUint64(blockNumber))
+  if err != nil{
+    return 0, err
+  }
+  return count, nil
+}
+
```

Comment:
Why does the linter not complaint about these spaces.

Reply:
The linter, at least for the pre-commit hook, has been broken for a while and mandates that all files in the repo are properly linted, not just what's in the commit. 
I spent some time trying to fix this during the retreat without luck. I think this has led to most people skipping the pre-commit hook, which isn't an issue because their IDE's do properly format the text.
Because I've had to be skipping the pre-commit hook recently as well to get things committed, I've been forgetting to do the lint sometimes and it's been ending up in the files.